README.md
# Octopress Date Format

Put nicely formatted dates on any post or page.

[![Build Status](https://travis-ci.org/octopress/date-format.svg)](https://travis-ci.org/octopress/date-format)
[![Gem Version](http://img.shields.io/gem/v/octopress-date-format.svg)](https://rubygems.org/gems/octopress-date-format)
[![License](http://img.shields.io/:license-mit-blue.svg)](http://octopress.mit-license.org)

## Installation

Add this line to your application's Gemfile:

    gem 'octopress-date-format'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install octopress-date-format

Next add it to your gems list in Jekyll's `_config.yml`

    gems:
      - octopress-date-format
    
## Usage

Any post (or page with a date) will automatically have some new date attributes. 

- `date` - The date, `2014-07-03 14:08:00 +0000`
- `date_text` - The formatted date, Jul 3rd, 2014
- `time_text` - The formatted time, 2:08 pm
- `date_xml` - The XML schema formatted date, 2014-07-03T14:08:00+00:00
- `date_html` - The formatted date in a `<time>` tag.
- `date_time_html` - The formatted date and time in a `<time>` tag.

Here's an example of what would be rendered with `{{ post.date_time_html }}`.

```html
<time class='entry-date' datetime='2014-07-03T14:08:00+00:00'>
  <span class='date'>
    <span class='date-month'>Jul</span>
    <span class='date-day'>3</span><span class='date-suffix'>rd</span>,
    <span class='date-year'>2014</span>
  </span>
  <span class='time'>2:08 pm</span>
</time>
```

If you update a post or page and want to display this in a template, add this to YAML
front-matter, `date_updated: 2014-07-03 15:03:15` and your post will have updated date
attributes as well.

- `updated` - The date, `2014-07-03 15:03:15 +0000`
- `date_updated_text` - The formatted date, Jul 3rd, 2014
- `time_updated_text` - The formatted time, 3:03 pm
- `date_updated_xml` - The XML schema formatted date, 2014-07-03T15:03:15+00:00
- `date_updated_html` - The formatted date in a `<time>` tag.
- `date_time_updated_html` - The formatted date and time in a `<time>` tag.

Of course you can use Liquid conditionals when outputting an updated date.

```
{% if post.updated %}Updated: {{ post.date_time_updated_html }}{% endif %}
```

## Configuration

You may change the formatting of the dates and times in
Jekyll's configuration file. Here are the defaults:

```
date_format: 'ordinal'     # July 3rd, 2014
time_format: '%-I:%M %P'   # 2:08 pm
```

To choose a different format, use a [Ruby strftime](http://apidock.com/ruby/DateTime/strftime)
compatible string. For example:

```
date_format: "%Y-%m-%d"  # e.g. 2014-07-03
time_format: "%H:%M"     # 24 hour time
```

## Contributing

1. Fork it ( https://github.com/octopress/date-format/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


## License

Copyright (c) 2013 Brandon Mathis

MIT License

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
# Changelog

### 3.0.3 - 2015-05-12

- Support for Jekyll 3.0 and Jekyll Hooks

### 3.0.2 - 2015-01-06

- Updated Octopress docs integration.

### 3.0.1 - 2014-10-12

- Documentation now works with Octopress docs.

### 3.0.0 - 2014-10-09

- Now only reads from Jekyll's configuration.

### 2.0.2 - 2014-10-09

- Also reads from Jekyll's configuration for broader compatibility.

### 2.0.1 - 2014-08-18

- Added support for old `updated` YAML front-matter var.

### 2.0.0 - 2014-08-03

- Initial re-release, used to be jekyll-date-format.


test/_plugins/jekyll-date-format.rb
require 'octopress-date-format'


lib/jekyll-date-format.rb

module Jekyll
  class DateFormat

    # Returns a datetime if the input is a string
    def self.datetime(date)
      if date.class == String
        date = Time.parse(date)
      end
      date
    end

    # Returns an ordidinal date eg July 22 2007 -> July 22nd 2007
    def self.ordinalize(date)
      date = datetime(date)
      "#{date.strftime('%b')} #{ordinal(date.strftime('%e').to_i)}, #{date.strftime('%Y')}"
    end

    def self.time_tag(date, formatted, type)
      t = "<time class='date-#{type}' datetime='#{date}'"
      t += " pubdate" unless type == 'updated'
      t += ">#{formatted}</time>"
    end


    # Returns an ordinal number. 13 -> 13th, 21 -> 21st etc.
    def self.ordinal(number)
      if (11..13).include?(number.to_i % 100)
        "#{number}<sup>th</sup>"
      else
        case number.to_i % 10
        when 1; "#{number}<sup>st</sup>"
        when 2; "#{number}<sup>nd</sup>"
        when 3; "#{number}<sup>rd</sup>"
        else    "#{number}<sup>th</sup>"
        end
      end
    end

    # Formats date either as ordinal or by given date format
    # Adds %o as ordinal representation of the day
    def self.format_date(date, format)
      date = datetime(date)
      if format.nil? || format.empty? || format == "ordinal"
        date_formatted = ordinalize(date)
      else
        date_formatted = date.strftime(format)
        date_formatted.gsub!(/%o/, ordinal(date.strftime('%e').to_i))
      end
      date_formatted
    end

  end

  class Post

    # Copy the #initialize method to #old_initialize, so we can redefine #initialize
    #
    alias_method :old_initialize, :initialize
    attr_accessor :updated

    def initialize(site, source, dir, name)
      old_initialize(site, source, dir, name)
      format = self.site.config['date_format']

      date_f = DateFormat.format_date(self.date, format) unless self.data['date'].nil?
      date_t = DateFormat.time_tag(self.date, date_f, 'published')

      self.data['date_formatted'] = date_f
      self.data['time_tag'] = date_t

      unless self.data['updated'].nil?
        updated = Time.parse(self.data['updated'].to_s)
        updated_f = DateFormat.format_date(self.data['updated'], format)
        updated_t = DateFormat.time_tag(self.date, date_f, 'updated')

        self.data['updated'] = updated
        self.data['updated_formatted'] = updated_f
        self.data['updated_time_tag'] = updated_t
      end
    end
  end

  class Page

    # Copy the #initialize method to #old_initialize, so we can redefine #initialize
    #
    alias_method :old_initialize, :initialize
    attr_accessor :updated, :date

    def initialize(site, source, dir, name)
      old_initialize(site, source, dir, name)
      format = self.site.config['date_format']

      unless self.data['date'].nil?

        date_f = DateFormat.format_date(self.data['date'], format) unless self.data['date'].nil?
        date_t = DateFormat.time_tag(self.data['date'], date_f, 'published')

        self.data['date_formatted'] = date_f
        self.data['time_tag'] = date_t
      end

      unless self.data['updated'].nil?
        updated = Time.parse(self.data['updated'].to_s)
        updated_f = DateFormat.format_date(self.data['updated'], format)
        updated_t = DateFormat.time_tag(self.date, date_f, 'updated')

        self.data['updated'] = updated
        self.data['updated_formatted'] = updated_f
        self.data['updated_time_tag'] = updated_t
      end
    end
  end
end



lib/octopress-date-format/configuration.rb
module Octopress
  unless defined? Octopress.config
    def self.config
      file = '_octopress.yml'
      if File.exist?(file)
        SafeYAML.load_file(file) || {}
      else
        {}
      end
    end
  end

  module PageDate
    module Configuration
      @jekyll_config = {}

      DEFAULTS = {
        'date_format' => 'ordinal',
        'time_format' => '%-I:%M %P'
      }

      def self.jekyll_config=(config)
        @jekyll_config = config
      end

      # For broader compatibility, also read config from Jekyll
      def self.jekyll_date_config
        @jekyll_config.select { |key,_| DEFAULTS.include? key }
      end
    end

    def self.config
      @config ||= begin
        default = PageDate::Configuration::DEFAULTS
        jekyll = PageDate::Configuration::jekyll_date_config
        octopress = Octopress.config

        config = Jekyll::Utils.deep_merge_hashes(default, jekyll)
        Jekyll::Utils.deep_merge_hashes(config, octopress)
      end
    end
  end
end



lib/octopress-date-format.rb
require 'octopress-date-format/version'

module Octopress
  module DateFormat

    class << self
      DEFAULTS = {
        'date_format' => 'ordinal',
        'time_format' => '%-I:%M %P'
      }
      attr_accessor :config

      def config=(config)
        @config = DEFAULTS.merge(config)
      end

      def hack_date(page)
        if page.data['date'] || page.respond_to?(:date)
          date = datetime(page.data['date'] || page.date)

          page.data['date_xml']  = date.xmlschema
          page.data['date_text'] = format_date(date)
          page.data['time_text'] = format_time(date)
          page.data['date_html'] = date_html(date, false)
          page.data['date_time_html'] = date_html(date)
        end

        # Legacy support
        if page.data['updated']
          page.data['date_updated'] = page.data['updated']
        end

        if page.data['date_updated']
          updated  = datetime(page.data['date_updated'])
          page.data['date_updated_xml']  = updated.xmlschema
          page.data['date_updated_text'] = format_date(updated)
          page.data['time_updated_text'] = format_time(updated)
          page.data['date_updated_html'] = date_updated_html(updated, false)
          page.data['date_time_updated_html'] = date_updated_html(updated)
        end
        page
      end

      def date_html(date, time=true)
        tag =  "<time class='entry-date' datetime='#{ date.xmlschema }'>"
        tag += "<span class='date'>#{format_date(date, true)}</span>"
        if time
          tag += " <span class='time'>#{format_time(date)}</span>" if time
        end
        tag += "</time>"
      end

      def date_updated_html(date, time=true)
        date_html(date, time).sub('entry-date','updated')
      end

      def format_date(date, html=false)
        format = self.config['date_format']
        if format == 'ordinal'
          html ? ordinalize_html(date) : ordinalize(date)
        else
          date.strftime(format)
        end
      end

      def format_time(date, html=false)
        format = config['time_format']
        date.strftime(format)
      end

      # Returns an ordidinal date eg July 22 2007 -> July 22nd 2007
      def ordinalize(date)
        "#{date.strftime('%b %-d')}#{ordinal_suffix(date)}, #{date.strftime('%Y')}"
      end

      def ordinalize_html(date)
        d = "<span class='date-month'>#{date.strftime('%b')}</span> "
        d += "<span class='date-day'>#{date.strftime('%-d')}</span>"
        d += "<span class='date-suffix'>#{ordinal_suffix(date)}</span>, "
        d += "<span class='date-year'>#{date.strftime('%Y')}</span>"
      end

      # Returns an ordinal number. 13 -> 13th, 21 -> 21st etc.
      def ordinal_suffix(date)
        number = date.strftime('%e').to_i
        if (11..13).include?(number % 100)
          "th"
        else
          case number % 10
          when 1; "st"
          when 2; "nd"
          when 3; "rd"
          else    "th"
          end
        end
      end

      def datetime(input)
        case input
        when Time
          input
        when Date
          input
        when String
          Time.parse(input) rescue Time.at(input.to_i)
        when Numeric
          Time.at(input)
        else
          raise "Invalid Date:", "'#{input}' is not a valid datetime."
          exit(1)
        end
      end
    end

    if defined?(Jekyll::Hooks)
      Jekyll::Hooks.register :site, :after_reset do |site|
        DateFormat.config = site.config
      end

      Jekyll::Hooks.register [:page, :post], :post_init do |item|
        DateFormat.hack_date(item)
      end
    else
      require 'octopress-hooks'

      class PageHook < Hooks::Page
        def post_init(page)
          DateFormat.hack_date(page)
        end
      end

      class PostHook < Hooks::Post
        def post_init(post)
          DateFormat.hack_date(post)
        end
      end

      class SiteHook < Hooks::Site
        def pre_read(site)
          DateFormat.config = site.config
        end
      end
    end
  end
end

if defined? Octopress::Docs
  Octopress::Docs.add({
    name:        "Octopress Date Format",
    gem:         "octopress-date-format",
    version:     Octopress::DateFormat::VERSION,
    description: "Put nicely formatted dates on any post or page",
    path:        File.expand_path(File.join(File.dirname(__FILE__), "../")),
    source_url:  "https://github.com/octopress/date-format",
  })
end
